### ISSUE TITLE
Projects need to depend on Apache Spark's `spark-core` to import `geotrellis-spark` types

### ISSUE THREAD

**Issue by fosskers** [locationtech/geotrellis#2081]
This has happened to me a few times when creating new projects that depend on GT. The error you get at compile time is quite cryptic, and it was only through trial-and-error that we discovered that adding the `spark-core` dep fixes it:

```
...
 [error] Could not access type RDD in package org.apache.spark.rdd,
 [error] because it (or its dependencies) are missing. Check your build definition for
 [error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
 [error] A full rebuild may help if 'Implicits.class' was compiled against an incompatible version of org.apache.spark.rdd.
 [error] missing or invalid dependency detected while loading class file 'Implicits.class'.
 [error] Could not access type RDD in package org.apache.spark.rdd,
 [error] because it (or its dependencies) are missing. Check your build definition for
 [error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
 [error] A full rebuild may help if 'Implicits.class' was compiled against an incompatible version of org.apache.spark.rdd.
 [error] missing or invalid dependency detected while loading class file 'package.class'.
 [error] Could not access type RDD in package org.apache.spark.rdd,
 [error] because it (or its dependencies) are missing. Check your build definition for
 [error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
 [error] A full rebuild may help if 'package.class' was compiled against an incompatible version of org.apache.spark.rdd.
 [error] missing or invalid dependency detected while loading class file 'package.class'.
 [error] Could not access type Partitioner in package org.apache.spark,
 [error] because it (or its dependencies) are missing. Check your build definition for
 [error] missing or conflicting dependencies. (Re-run with `-Ylog-classpath` to see the problematic classpath.)
 [error] A full rebuild may help if 'package.class' was compiled against an incompatible version of org.apache.spark.
...
```

Intuitively we might think "of course a user would need `spark-core` if they're doing things with `geotrellis-spark`", except that there are use cases where only the types (i.e. `SpatialKey`) are useful, with no Spark involved at all.

Suggestions:

1. Dive into GT config to try and resolve this w/o changing the API, or;
2. Move many of the basic GT types out of the `spark` subpackage into a `geotrellis-core` package.

**Comment by pomadchin** [comment]
If you'll launch your assembly without explicit `spark` or `hadoop` dep via `spark-submit` there would be no such issues.

That's because of the `"provided"` flag for [hadoop and spark deps](https://github.com/locationtech/geotrellis/blob/master/spark/build.sbt#L5-L6). I think it's a correct thing. By that flag we allow clients to choose what `Spark` or `Hadoop` version to use (`Apache Spark` and `Apache Hadoop` have a strong binary compatibility with previous versions) and to avoid lots of `excludes` in client `build.sbt` code (that would happen in 99% cases, because these artifacts are useless in a spark job).

**Comment by fosskers** [comment]
This doesn't even involve `spark-submit`, since the side project that depends on `geotrellis-spark` won't even *compile* unless `spark-core` is added as well.

**Comment by pomadchin** [comment]
That's fine, that's the way to avoid `excludes` like [these](https://github.com/locationtech/geotrellis/blob/master/geowave/build.sbt#L10)

**Comment by pomadchin** [comment]
I think a note in docs would be enough. It's a common solution like in [GeoMesa](https://github.com/locationtech/geomesa/blob/master/geomesa-spark/geomesa-spark-core/pom.xml#L17).

**Comment by lossyrob** [comment]
I think the deeper issue here is that we have some types that don't need spark to be useful, but exist in the spark subproject. All the tiling logic, the value readers, collections api are not spark specific, so by default having people rely on that project that contains spark dependencies seems off and could be corrected in 2.0

**Comment by metasim** [comment]
Using [this trick here](http://stackoverflow.com/a/21803413/296509) might be worth trying to see if it fixes the issue. It at least allows you to use `sbt run` on main classes and ensure the `provided` libraries are available. Regardless, I'm definitely in the camp that things developer ergonomics on this are extremely important. Nothing puts a new user off faster than some obtuse compiler error in their first `GTHelloWorld`.

**Comment by pomadchin** [comment]
Resolved in 3.0
